R/dv8.R

```diff
@@ -0,0 +1,228 @@
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#' Transforms a DV8 binary cluster file to a json cluster file.
+#'
+#' Converts a clustering binary in the dv8 format *-clsx.dv8-cslx into a
+#'  *-clsx.json.
+#'
+#' @param dv8_path path to dv8 binary
+#' @param clsxb_path path to clsx to convert
+#' @param clsxj_path path to save output file
+#' @export
+#' @family dv8
+dv8_clsxb_to_clsxj <- function(dv8_path, clsxb_path, clsxj_path){
+  system2(dv8_path, args=c('core:export-cluster', '-outputFile', clsxj_path, clsxb_path), stdout=FALSE, stderr=FALSE)
+}
+
+#' Parse a DV8 json cluster file to R data.table.
+#'
+#' Parses a *-clsx.json into a data table.
+#'
+#' @param clsxj_path path to clustering json (obtained from dv8_clsxb_to_clsxj)
+#' @export
+#' @family parsers
+#' @family dv8
+parse_dv8_clusters <- function(clsxj_path){
+  # Read in the clustering file
+  dv8_clusters_json <- jsonlite::read_json(clsxj_path)
+
+  # Get number of cluster layers in the json
+  num_layers <- length(dv8_clusters_json$structure)
+
+  # The structure field holds the relevant data
+  structure <- dv8_clusters_json[["structure"]]
+
+  # Will hold each row (a list of the current layer, module, and file) for the data table
+  cluster_list <- list()
+
+  # Iterate through each cluster
+  for (i in 1: num_layers) {
+    # Get the layer number (L0, L1, L2, ...)
+    layer_i <- structure[[i]]$name
+
+    # Get the number of modules for the current layer
+    num_modules <- length(structure[[i]]$nested)
+
+    # Iterate through each module
+    for (j in 1: num_modules) {
+      # Get the module number (M0, M1, M2, ...)
+      LM_name <- stringr::str_split(structure[[i]]$nested[[j]]$name, "/")
+      module_j <- LM_name[[1]][2]
+
+      # Get number of files for the layer/module pair
+      num_files_per_LM <- length(structure[[i]]$nested[[j]]$nested)
+
+      # Iterate through each file
+      for (k in 1: num_files_per_LM) {
+        # Get the file name
+        filename_k <- structure[[i]]$nested[[j]]$nested[[k]]$name
+
+        # Represents a row in the table with the layer #, module #, and filename
+        row_ijk <- list(filename_k, module_j, layer_i)
+
+        # Append current row (made up of the layer number, module number, & filename) to a list
+        cluster_list <- append(cluster_list, list(row_ijk))
+      }
+    }
+  }
+
+  # Create table from list of layer,module,file rows
+  cluster_parsed <- data.table::rbindlist(cluster_list)
+
+  # Rename result columns
+  cluster_parsed <- data.table::setnames(cluster_parsed, c("V1", "V2", "V3"), c("filename", "module", "layer"))
+  # Return parsed cluster data table
+  return(cluster_parsed)
+}
+
+#' Convert dependencies table to a sdsm.json
+#'
+#' Converts table of dependencies from Depends into an *-sdsm.json binary.
+#'
+#' @param depends_table parsed dependencies table
+#' @param project_name project name for the sdsm.json
```

Comment:
Hmm... I may be contradicting earlier feedback here, but we may want to call this variable sdmj_path for consistency here. 

The Notebook, however, should handle the logic of obtaining the project name out of the project configuration file, no? Please let me know if this conflicts to something we discussed prior.

Reply:
> which variable would be sdmj_path? depends_table? I thought we would take as input the table object from parse_dependencies and not a filepath?

No, `project_name`.